test: exempt previous release binaries from valgrind

Sjors opened the pull request:
Some, but not all, backward compatibility tests fail for me and it seems useless to run old release binaries under valgrind anyway.

Can be tested by running `test/functional/feature_txindex_compatibility.py --valgrind --timeout-factor=10` with and without this PR.

—
The previous version of this PR disabled these test entirely under valgrind. The current version does run the test, but starts the old binaries without valgrind.

Sjors commented:
Example failure:

```
```
$ test/functional/feature_txindex_compatibility.py --valgrind --timeout-factor=4
2023-03-08T14:43:48.054000Z TestFramework (INFO): Initializing test directory /tmp/bitcoin_func_test_c2c_8427
2023-03-08T14:43:49.852000Z TestFramework (ERROR): Assertion failed
Traceback (most recent call last):
  File "/home/sjors/dev/bitcoin/test/functional/test_framework/test_framework.py", line 559, in start_nodes
    node.wait_for_rpc_connection()
  File "/home/sjors/dev/bitcoin/test/functional/test_framework/test_node.py", line 232, in wait_for_rpc_connection
    'bitcoind exited with status {} during initialization'.format(self.process.returncode)))
test_framework.test_node.FailedToStartError: [node 0] bitcoind exited with status 1 during initialization

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/sjors/dev/bitcoin/test/functional/test_framework/test_framework.py", line 133, in main
    self.setup()
  File "/home/sjors/dev/bitcoin/test/functional/test_framework/test_framework.py", line 297, in setup
    self.setup_network()
  File "test/functional/feature_txindex_compatibility.py", line 39, in setup_network
    self.start_nodes()
  File "/home/sjors/dev/bitcoin/test/functional/test_framework/test_framework.py", line 562, in start_nodes
    self.stop_nodes()
  File "/home/sjors/dev/bitcoin/test/functional/test_framework/test_framework.py", line 577, in stop_nodes
    node.stop_node(wait=wait, wait_until_stopped=False)
  File "/home/sjors/dev/bitcoin/test/functional/test_framework/test_node.py", line 344, in stop_node
    self.stop()
  File "/home/sjors/dev/bitcoin/test/functional/test_framework/test_node.py", line 190, in __getattr__
    assert self.rpc_connected and self.rpc is not None, self._node_msg("Error: no RPC connection")
AssertionError: [node 0] Error: no RPC connection
2023-03-08T14:43:49.903000Z TestFramework (INFO): Stopping nodes
Traceback (most recent call last):
  File "test/functional/feature_txindex_compatibility.py", line 91, in <module>
    TxindexCompatibilityTest().main()
  File "/home/sjors/dev/bitcoin/test/functional/test_framework/test_framework.py", line 157, in main
    exit_code = self.shutdown()
  File "/home/sjors/dev/bitcoin/test/functional/test_framework/test_framework.py", line 313, in shutdown
    self.stop_nodes()
  File "/home/sjors/dev/bitcoin/test/functional/test_framework/test_framework.py", line 577, in stop_nodes
    node.stop_node(wait=wait, wait_until_stopped=False)
  File "/home/sjors/dev/bitcoin/test/functional/test_framework/test_node.py", line 344, in stop_node
    self.stop()
  File "/home/sjors/dev/bitcoin/test/functional/test_framework/test_node.py", line 190, in __getattr__
    assert self.rpc_connected and self.rpc is not None, self._node_msg("Error: no RPC connection")
AssertionError: [node 0] Error: no RPC connection
[node 2] Cleaning up leftover process
[node 1] Cleaning up leftover process
[node 0] Cleaning up leftover process
```

fanquake commented:
This seems arbitrary, and the `can't run this test under valgrind` message is not useful. I assume the first questions someone will have are going to be "Why can't I run _this_ test under Valgrind?", "Does it need fixing then?" etc.

Sjors commented:
I disabled all tests that use previous releases under valgrind. Also tried a different error message.

> "Does it need fixing then?" 

I added a code comment and some more context to the commit to help answer that, if anyone _does_ want to run these under valgrind.

Sjors commented on test/functional/feature_txindex_compatibility.py:
```diff
@@ -25,6 +25,8 @@ def set_test_params(self):
 
     def skip_test_if_missing_module(self):
         self.skip_if_no_previous_releases()
+        # Not all release binaries work with valgrind.
```
Thought about that, but there may other (future) tests we want to skip under valgrind (e.g. things that use a brittle dependency). Conversely, since some of tests _do_ work, maybe someone want to enable one later.

fanquake replied:
>  (future) tests we want to skip under valgrind

Let's not worry about unwritten tests, that we apparently might not want to test properly for some arbitrary reason.

Sjors replied:
I was also thinking existing tests, e.g. I haven't been able to run valgrind with BDB configured (didn't try very hard).

fanquake replied:
> Which raises the question how many settings we want --valgrind to be supported under.

I thnk currently the only realistic place we can fully support Valgrind, is within the CI system, where things like BDB are working, and we can more easily control settings and suppressions. Outside of that, things may work fine, or may not at all, depending on the system where it is being run.

luke-jr reviewed (changes requested):
Approach NACK: Would prefer to run all the tests, but only execute the older binaries without Valgrind (ie, still run the current version with Valgrind)

Sjors commented:
Running the old binaries without valgrind makes sense. I'll see if I can figure out how.

Sjors committed:
test: don't run old binaries under valgrind

This is unnecessary and caused test failures. The backward
compatibility tests are meant to find regressions in the
current codebase, not to detect bugs in older releases.

Sjors commented:
Changed this PR so that it now runs the old binaries without valgrind, rather than skipping the backward compatibility test entirely.

Sjors commented:
Rebased.
